### PR TITLE
mgr/dashboard_v2: Add frontend login/logout

### PR DESCRIPTION
**Pull request by ricardoasmarques**
This PR adds the frontend code for login and logout pages, including an interceptor to redirect unauthorized users to the login page, and display notifications when HTTP error codes are received.

screenshot from 2018-01-25 16-26-06

Signed-off-by: Ricardo Marques <rimarques@suse.com>

### CODE REVIEW

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/empty/empty.component.html [hunk 1]
```diff
@@ -0,0 +1 @@
+<!-- Empty Component, delete me! -->
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/shared/empty/empty.component.html]: Which purpose serves this empty component?

**Review by Devp00l — APPROVED** [review]
Awesome :+1:

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/empty/empty.component.html (hunk 1 above)
Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/shared/empty/empty.component.html]: Should be deleted, in the future, by the developer that implements the first route / page.

**Comment by tspmelo** [comment]
Since you inserted the navigation rules in openattic-theme.scss, and the ones in `navigation.component.scss` aren't being applied correctly in the navigation component, I would suggest you removed the rules from `navigation.component.scss`.

**Comment by ricardoasmarques** [comment]
@tspmelo Done

**Comment by ImTheKai** [comment]
lgtm

**Review by ImTheKai — APPROVED** [review]

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { Injectable } from '@angular/core';
+
+@Injectable()
+export class AuthService {
+
+  set(user) {
+    localStorage.setItem('oa_user', user);
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts]: Rename `oa_` prefix to something else

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  }
+
+  remove() {
+    localStorage.removeItem('oa_user');
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts]: same as above

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  }
+
+  isLoggedIn() {
+    return localStorage.getItem('oa_user') !== null;
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts]: same as above

File: src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  ngOnInit() {
+    if (this.authService.isLoggedIn()) {
+      this.router.navigate(['']);
+    }
+  }
+
+  login() {
+    this.http.post(`/api/auth`, this.model).subscribe((data: any) => {
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.ts]: I believe the best practices say that the HTTP request should be done within the `AuthService` code.
For example add a method to `AuthService` similar to this:

```
import 'rxjs/add/operator/toPromise';

/* ... */

login(credentials: Login): Promise<any> {
    return this.http.post('/api/auth', credentials).toPromise().then(resp => {
        let data = resp.json();
        this.set(data.username);
    })
}
```

Then here you can call the service like this:

```
this.authService.login(this.model).then(resp => {
  this.router.navigate([''])
});
```

File: src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.model.ts [hunk 6]
```diff
@@ -0,0 +1,4 @@
+export class Login {
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.model.ts]: I would call it `Credentials`

File: src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/logout/logout.component.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+  constructor(private authService: AuthService,
+              private router: Router,
+              private http: HttpClient) { }
+
+  ngOnInit() {
+  }
+
+  logout() {
+    this.http.delete('/api/auth').subscribe(() => {
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/logout/logout.component.ts]: Same comment as in the `login.component.ts`, the HTTP request should be done from the `AuthService` code,

**Review by rjfd — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.html [hunk 8]
```diff
@@ -0,0 +1,52 @@
+<div class="login">
+  <div class="row full-height vertical-align">
+    <div class="col-sm-6 hidden-xs">
+      <img src="assets/Ceph_Logo_Stacked_RGB_White_120411_fa_256x256.png"
+           alt="openATTIC"
```
Comment (LenzGr) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.html]: s/openATTIC/Ceph Logo/

File: src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
+import { Component, OnInit, ViewContainerRef } from '@angular/core';
+import { Login } from './login.model';
+import { AuthService } from '../../../shared/services/auth.service';
+import { Router } from '@angular/router';
+import { HttpClient } from '@angular/common/http';
+import { ToastsManager } from 'ng2-toastr';
+
+@Component({
+  selector: 'oa-login',
```
Comment (LenzGr) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.ts]: s/oa-login/dashboard-login/ maybe?

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/empty/empty.component.html (hunk 1 above)
Comment (LenzGr) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/shared/empty/empty.component.html]: How about adding this information to the comment then?

**Review by LenzGr — APPROVED** [review]

File: src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        LoginComponent
+      ]
+    })
+    .compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(LoginComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
```
Comment (votdev) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.spec.ts:33]: Can be replaced by using the ComponentFixtureAutoDetect provider.

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts [hunk 11]
```diff
@@ -0,0 +1,18 @@
+import { Injectable } from '@angular/core';
+
+@Injectable()
+export class AuthService {
+
+  set(user) {
```
Comment (votdev) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts]: Is user a string? Maybe we can use ``user: string`` here?

**Review by votdev — APPROVED** [review]

File: src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.spec.ts (hunk 10 above)
Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.spec.ts:33]: This was generated by the "ng generate component" so I will not change it at the moment, to guarantee that we have the same structure on all tests.

We can change it later when we stop to focus on the frontend unit tests.

File: src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.ts (hunk 5 above)
Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.ts]: Done

File: src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.model.ts (hunk 6 above)
Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.model.ts]: Done

File: src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/logout/logout.component.ts (hunk 7 above)
Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/logout/logout.component.ts]: Done

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts (hunk 2 above)
Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts]: Done

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts (hunk 3 above)
Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts]: Done

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts (hunk 4 above)
Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts]: Done

File: src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.html (hunk 8 above)
Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.html]: Done

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts (hunk 11 above)
Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/shared/services/auth.service.ts]: Done

**Review by rjfd — APPROVED** [review]
lgtm

File: src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.ts (hunk 9 above)
Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/core/auth/login/login.component.ts]: Angular components prefix renamed from 'oa' to 'cd'

**Pull request merged**